cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java

```diff
@@ -207,48 +258,49 @@ public CloseableIterator<StructuredRow> scan(Range keyRange, int limit, Field<?>
     if (!schema.isIndexColumn(filterIndex.getName())) {
       throw new InvalidFieldException(schema.getTableId(), filterIndex.getName(), "is not an indexed column");
     }
-    ScannerIterator scannerIterator = new ScannerIterator(getScanner(keyRange), schema);
-    FilterByIndexIterator filterByIndexIterator = new FilterByIndexIterator(scannerIterator, filterIndex, schema);
+    FilterByFieldIterator filterByIndexIterator =
+      new FilterByFieldIterator(getFilterByRangeIterator(keyRange), filterIndex, schema);
     return new LimitIterator(Collections.singleton(filterByIndexIterator).iterator(), limit);
   }
 
+  /*
+   * Scan the nosql DB with provided ranges, find out the most outbound range, scan and filter by ranges,
```

Comment:
This looks like a significant change. Instead of doing multiscan, you actually do a single scan with postfiltering. If keyRanges are far away this would mean a very long scan. Why are you doing it and how are you sure it's not going to cost a lot

Reply:
YES exactly, this method is actually doing a single scan on the most outbound range. In our real cases, the keyRanges won't be far away. And even if so, it's running nosql on local sandbox, we agreed upon the potential performance degrade

Reply:
"This won't be too bad in real case since the actual passed in ranges usually have one outbound. " - this is not true. Into io.cdap.cdap.gateway.handlers.ProgramLifecycleHttpHandler#getStatuses one can pass programs from different namespaces that would give a horrible scan

Reply:
They can be far away. e.g. RUN_RECORDS_SPEC field field is RUN_STATUS, so scanning for multiple run statuses gives you a big scan. In other cases it's often namespace, so any call for multiple namespaces would given you a big scan. If we are talking about local sandbox only, then old implementation would work as well, no need to change.

If you really want to optimize here, you can relax the merging. bit - allow merging when at least first field overlaps (e.g. same namespace). This would ensure there are no wild scans across half a table

Reply:
Also this would solve your use case - scanning a number of runs for program(s) in same namespace with same run status.

Reply:
So the previous implementation I had, which does scans on the ranges one by one, post filtering and deduplicate and then sorting. This could potentially scan on the same range or similar ranges multiple times, also potentially cost more memories and time

Reply:
Yes. But new implementation of collapsing everything to a single range can potentially perform very wild scans. That's why I'd propose to collapse ranges only if they share at least 1 field an scan them separately otherwise. This would solve the "multiple namespaces" or "multiple run statuses" use case, but would still collapse into a single scan most of the time

Reply:
I'll put a TODO ticket here, while we're considering the pros and cons of option (A) One single scan but potentially resulting in wild scan v.s. (B) Actual multiscans which might have repeated results

fyi  @tivv @chtyim @yaojiefeng

Reply:
Actually I'll combine these two options, merge the longest prefix keys, scan these intervals, post filtering. Still, will do it in the TODO

Reply:
I am fine with a follow-up PR given it's done in 6.8.0. Please raise a JIRA for 6.8.0 and refer in TODO

Reply:
I've refactored the code with latest commit